### DISCUSSION TITLE
Native build test with gRPC client

### DISCUSSION THREAD

**Discussion by DSkoufis** [quarkusio/quarkus#43679] [Q&A] [answered]
Are there any examples about how we can add an Integration test for a native executable that uses a gRPC service? The server is external to the app under test.

In our `@QuarkusTest` we're using a `@MockBean` annotation so it's all mocked but I'm aware that we can't use the same in ITs.

I've found something about an `in-process` builder in the gRPC plugin but not sure how to use it and where to begin with it.

[a comment by an automated account is not shown]

**Reply by cescoffier** [reply]
I would not go with the in-process, but instead implement a version of the service in your test.

**Reply by DSkoufis** [reply]
Ah! As in a service implementation that will be available only in the tests and start when native tests are about to start? That makes sense.

Any hints how that can be supported? Is there a quarkus annotation/resource that allows us to start the server or is there a way to hook that startup with the failsafe plugin (that starts the native exe)?

**Reply by cescoffier** [reply, marked as answer]
You can start the Vert.x gRPC server in your tests. The only issue is that you will need to use a fixed port.
Here is an example to start a gRPC server with a HelloService:
https://vertx.io/docs/vertx-grpc/java/#_service_bridge

**Reply by DSkoufis** [reply]
Many thanks for the replies. I had a look at the service bridge but in the example they have it's like:

```
GrpcIoServer grpcServer = GrpcIoServer.server(vertx);
```

Is there a way to retrieve the Vert.x instance in the integration tests? Or do I have to create my own in the tests?

> The only issue is that you will need to use a fixed port

About that, if I understand what you suggest correctly, can't we use a random port through a `QuarkusTestResourceLifecycleManager` instance?

**Reply by cescoffier** [reply]
You can create a new Vertx instance with: Vertx.vertx();

**Reply by DSkoufis** [reply]
Ah so there’s no need to use one from Quarkus’ context. Brilliant that’s really helpful. I’ll mark the previous reply as the answer as it’s giving the starting point for the solution.

Thank you
